server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java

```diff
@@ -414,6 +414,11 @@ public void cleanFiles(int totalTranslogOps, Store.MetadataSnapshot sourceMetaDa
                 indexShard.shardPath().resolveTranslog(), SequenceNumbers.UNASSIGNED_SEQ_NO, shardId,
                 indexShard.getPendingPrimaryTerm());
             store.associateIndexWithNewTranslog(translogUUID);
+
+            assert indexShard.getRetentionLeases().leases().isEmpty() : indexShard.getRetentionLeases(); // not loaded yet
```

Comment:
peer recoveries can be retried if they fail midway through, so I wonder if this can be violated

Reply:
I tried adding some recovery failures in 099005816fc98e1b5e37cd2d171d38e806f70cb8 but this didn't trigger this assertion. But then I tried failing recoveries the other way round too in 8391e5677c2fb1a14bd79ad47ac3a544444fb676 and did indeed get this to trigger, so have adjusted this code to deal with this.